packages/teleport-generator-shared/src/typings/generators.ts

```diff
@@ -187,11 +187,71 @@ export interface ProjectGeneratorOutput {
 
 export type GenerateProjectFunction = (
   input: Record<string, unknown>,
-  options: ProjectGeneratorOptions
+  template: TemplateDefinition,
+  options?: ProjectGeneratorOptions
 ) => Promise<ProjectGeneratorOutput>
 
 export type GenerateComponentFunction = (
   // TODO rename to ComponentGeneratorOptions
   input: Record<string, unknown>,
   options?: GeneratorOptions
 ) => Promise<CompiledComponent>
+
+/**
+ * Interfaces used in the publishers
+ */
+export type PublisherFactory<T, U> = (configuration?: Partial<T & PublisherFactoryParams>) => U
+
+export interface Publisher<T, U> {
+  publish: (options?: T) => Promise<PublisherResponse<U>>
+  getProject: () => GeneratedFolder | undefined
+  setProject: (project: GeneratedFolder) => unknown
+}
+
+export interface PublisherFactoryParams {
+  project?: GeneratedFolder
+  projectName?: string
+}
+export interface PublisherResponse<T> {
+  success: boolean
+  payload?: T
+}
+
+/**
+ * Interfaces used in the packers
+ */
+export interface AssetsDefinition {
+  assets: AssetInfo[]
+  meta?: {
+    prefix: string | string[]
+  }
+}
+
+export interface AssetInfo {
+  data: string
+  name: string
+  type: string
+}
+
+export interface TemplateDefinition {
+  templateFolder?: GeneratedFolder
+  remote: RemoteTemplateDefinition
+  meta?: {
+    componentsPath?: string[]
```

Comment:
would this work better as a Record<string, string[]> structure? this way you don't have to hardcode here all the possible values

Reply:
Record<string, string[]> would work here, but you won't have autocomplete when defining the template metadata. Let's have a detailed talk about this.